SourceCode/Scene/CRTMesh.cpp

```diff
@@ -1,19 +1,145 @@
 #include "CRTMesh.h"
 
-CRTMesh::CRTMesh(const std::vector<CRTVector>& vertices, const std::vector<int>& triangleVertIndices) : 
-    vertices(vertices), triangleVertIndices(triangleVertIndices)
-{}
+void CRTMesh::calculateFaceNormals() {
+    for (int i = 0; i < triangleVertIndices.size(); i += VERTICES) {
+        CRTVector v0 = vertices[triangleVertIndices[i + 0]];
+        CRTVector v1 = vertices[triangleVertIndices[i + 1]];
+        CRTVector v2 = vertices[triangleVertIndices[i + 2]];
+        CRTVector V1 = v1 - v0;
+        CRTVector V2 = v2 - v0;
 
-std::vector<CRTTriangle> CRTMesh::getTriangles() const
+        CRTVector normal = cross(V1, V2); // we set it to the counterclockwise cross-product
+        normal.normalize();
+        faceNormals[i / VERTICES] = std::move(normal);
+    }
+}
+
+void CRTMesh::calculateVertexNormals() {
+    for (int i = 0; i < vertices.size(); i++) {
+        CRTVector vNormal;
+        int triangleParticipations = 0;
+        for (int j = 0; j < triangleVertIndices.size(); j++) {
+            if (triangleVertIndices[j] == i) { // then a triangle uses the i-th vertex, we add its face normal to our calculation
+                int triangleIndex = j / VERTICES; // vertex indices 0 - 2 are from triangle 0, 3 - 5 from triangle 1 and so on
+                vNormal += faceNormals[triangleIndex];
+                triangleParticipations++; 
+            }
+        }
+        vNormal *= (1.0f / triangleParticipations); // since each face normal has len 1 we only need to scale it down by the number of vectors we've added
+        vertexNormals[i] = std::move(vNormal);
+    }
+}
+
+CRTMesh::CRTMesh(const std::vector<CRTVector>& vertices, const std::vector<int>& triangleVertIndices) : CRTMesh(vertices, triangleVertIndices, DEFAULT_MATERIAL)
+{
+}
+
+CRTMesh::CRTMesh(const std::vector<CRTVector>& vertices, const std::vector<int>& triangleVertIndices, const CRTMaterial& material) :
+    vertices(vertices), triangleVertIndices(triangleVertIndices), material(material)
+{
+    faceNormals.resize(triangleVertIndices.size() / VERTICES);
+    vertexNormals.resize(vertices.size());
+    calculateFaceNormals();
+    calculateVertexNormals();
+}
+
+CRTVector CRTMesh::calculateSmoothNormal(int triangleIndex, const CRTVector& point) const {
+    CRTTriangle triangle(vertices[triangleVertIndices[triangleIndex + 0]],
+        vertices[triangleVertIndices[triangleIndex + 1]],
+        vertices[triangleVertIndices[triangleIndex + 2]],
+        faceNormals[triangleIndex / VERTICES]);
+    CRTVector barycenticCoordinates = triangle.getBarycenticCoordinates(point);
```

Comment:
You could pass the bary coordinates directly using intersection data structure, with information for the hit.